content/post/2022-04-25-PHP-database-result-iteration.md

```diff
@@ -0,0 +1,164 @@
+
+- You're subject to the [PHP interpreter rules of what evaluates to `true` and `false`](https://www.php.net/manual/en/language.types.boolean.php#language.types.boolean.casting). Your code might still work for some cases, but not all.
+- You'll **override** `$a` with the value of `$b`
+
+In the case of looping over a database result, you want exactly this behavior, but in almost all other cases, using assignment instead of comparison is an unobtrusive and hard-to-find typo that makes your code behave in unexpected ways. To protect against this typical mistake, the authors of the [MediaWiki Coding Style](https://www.mediawiki.org/wiki/Manual:Coding_conventions/PHP) decided to disallow assignments in control structures.
+
+Even when you are using a different coding style, I can recommend using
+this rule in the PHP Cosing Style Checker (or an IDE setting that highlights problematic code pieces) to
+prevent severe errors that come from a simple typo.
+
+#### Side note: Yoda conditions
+Other coding styles like [Symfony](https://symfony.com/doc/current/contributing/code/standards.html#structure) and [WordPress](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#yoda-conditions) allow assignments in comparisons, but prevent accidental assignments in comparisons by enforcing a pattern called "[Yoda Conditions](https://en.wikipedia.org/wiki/Yoda_conditions)". This style of comparison places the static comparison value to the left instead of the right, triggering a syntax error when you accidentally assign:
+
+```php
+// The usual comparison
+if ($myVariable == 1) { /* ... */}
+
+// Yoda comparison
+if (1 == $myVariable) { /* ... */}
+
+// Syntax error when you accidentally assign
+if (1 = $myVariable) { /* ... */}
+
+```
+
+Personally, I don't like Yoda conditions, because they break the expected reading direction and don't prevent the override when the left side is a variable. And to enforce this style, you need a coding style checker anyway, so why not disallow assignments in control structures in your CS checker?
+
+
+### How to improve the database iteration code
+#### Explicit assignment outside condition
+
+When using the [MediaWiki Coding Style](https://www.mediawiki.org/wiki/Manual:Coding_conventions/PHP), the code that doesn't violate the rule looks like this:
+
+```php
+$row = $result->fetch();
+while($row !== null ) {
+    // do something with $row
+    // ...
+   $row = $result->fetch();
+}
+```
+
+While it satisfies the code style checker rules, the code becomes worse in two ways:
+
+- It duplicates the call to `fetch()`, violating the ["Don't repeat yourself" principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
+- It makes it easier to forget to call `fetch()` at the end of the loop, resulting in an infinite loop. This could be mitigated by using a `do ... while` loop.
```

Comment:
You’d also have to repeat the `fetch()` call before any `continue;`… I can’t think of a good way to fit that in here, though. Maybe it’s just not necessary to mention :)